### PR TITLE
[WIP] Xyk refactor

### PR DESCRIPTION
**Pull request by iStrike7**
To be merged with:
https://github.com/mangata-finance/substrate/pull/87
https://github.com/mangata-finance/moonbeam/pull/23
https://github.com/mangata-finance/cumulus/pull/18
https://github.com/mangata-finance/open-runtime-module-library/pull/28

### CODE REVIEW

File: pallets/proof-of-stake/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,916 @@
...
+			liquidity_asset_id,
+			&user,
+			liquidity_assets_burned,
+		);
+
+		Ok(())
+	}
+}
+
+impl<T: Config> XykRewardsApi for Pallet<T> {
```
Comment (mateuszaaa) [inline pallets/proof-of-stake/src/lib.rs:844]: i would rather get rid of `XYK` from the trait name, maybe sth like CumulativeWorkRewardsApi ?

**Review by mateuszaaa — APPROVED** [review]

File: pallets/proof-of-stake/src/lib.rs (hunk 1 above)
Comment (iStrike7) [reply pallets/proof-of-stake/src/lib.rs]: Done

File: pallets/proof-of-stake/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,913 @@
...
+			liquidity_asset_id: TokenId,
+			amount_permille: Permill,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+
+			<T::Xyk as XykFunctionsTrait<T::AccountId>>::do_compound_rewards(
+				sender,
+				liquidity_asset_id.into(),
+				amount_permille,
+			)?;
```
Comment (mateuszaaa) [inline pallets/proof-of-stake/src/lib.rs:173]: that seems more like ProofOfStakeRewards functionality than XYK to me :thinking:

File: pallets/proof-of-stake/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,913 @@
...
+				amount,
+			)?;
+
+			Ok(())
+		}
+
+		#[transactional]
+		#[pallet::call_index(2)]
+		#[pallet::weight(<<T as Config>::WeightInfo>::claim_rewards_v2())]
+		pub fn claim_rewards_all_v2(
```
Comment (mateuszaaa) [inline pallets/proof-of-stake/src/lib.rs:200]: can we remove v2 postfix

File: pallets/proof-of-stake/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,913 @@
...
+			<Self as ProofOfStakeRewardsApi<T::AccountId>>::update_pool_promotion(
+				liquidity_token_id,
+				Some(liquidity_mining_issuance_weight),
+			)
+		}
+
+		#[transactional]
+		#[pallet::call_index(4)]
+		#[pallet::weight(<<T as Config>::WeightInfo>::activate_liquidity_v2())]
+		pub fn activate_liquidity_v2(
```
Comment (mateuszaaa) [inline pallets/proof-of-stake/src/lib.rs:233]: can we remove v2 postfix ?

File: pallets/proof-of-stake/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,913 @@
...
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn native_token_id() -> TokenId {
+		<T as Config>::NativeCurrencyId::get()
+	}
+
+	fn calculate_rewards_v2(
```
Comment (mateuszaaa) [inline pallets/proof-of-stake/src/lib.rs:273]: remove v2 postfix

**Review by mateuszaaa — CHANGES_REQUESTED** [review]

**Comment by mateuszaaa** [comment]
Can we get rid of all v2 postfixes that does not require storage migrations ? :thinking:

File: pallets/proof-of-stake/src/lib.rs (hunk 2 above)
Comment (iStrike7) [reply pallets/proof-of-stake/src/lib.rs:173]: Implementation wise we would still prefer do_compound_rewards to be in Xyk itself.

File: pallets/proof-of-stake/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,913 @@
...
+		AccountIdOf<T>,
+		Twox64Concat,
+		TokenId,
+		RewardInfo,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn liquidity_mining_active_pool_v2)]
+	pub type LiquidityMiningActivePoolV2<T: Config> =
```
[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/proof-of-stake/src/lib.rs:155]: Agreed in slack to not do this rn

**Review by mateuszaaa — APPROVED** [review]
lgtm

**Review by majusko — DISMISSED** [review]

**Comment by iStrike7** [comment]
> Can we get rid of all v2 postfixes that does not require storage migrations ? thinking

As discussed on slack, we are not doing this to avoid api changes

**Review by mateuszaaa — DISMISSED** [review]
lgtm

**Review by mateuszaaa — APPROVED** [review]
lgtm

**Pull request merged**